cpp/src/arrow/util/rle_encoding_internal.h

```diff
@@ -299,385 +552,988 @@ class RleEncoder {
+    if (ARROW_PREDICT_FALSE(count == 0 || count > kMaxCount)) {
+      /// Illegal number of encoded values
+      return {};
     }
-    if (block.AllSet()) {
-      processed = GetBatchWithDict<T>(dictionary, dictionary_length, out, block.length);
-    } else if (block.NoneSet()) {
-      converter.FillZero(out, out + block.length);
-      processed = block.length;
-    } else {
-      processed = GetSpaced<T, /*RunType=*/IndexType, DictionaryConverter<T>>(
-          converter, block.length, block.length - block.popcount, valid_bits,
-          valid_bits_offset, out);
+
+    auto const values_count = static_cast<values_count_type>(count * 8);
+    ARROW_DCHECK_LT(count, internal::max_size_for_v<raw_data_size_type>);
+    // Count Already divided by 8
+    auto const bytes_read =
+        header_bytes + static_cast<raw_data_size_type>(count) * value_bit_width_;
+
+    auto control = handler.OnBitPackedRun(
+        BitPackedRun(data_ + header_bytes, values_count, value_bit_width_));
+
+    return {bytes_read, control};
+  }
+
+  using values_count_type = RleRun::values_count_type;
+  if (ARROW_PREDICT_FALSE(
+          count == 0 ||
+          count > static_cast<uint32_t>(std::numeric_limits<values_count_type>::max()))) {
+    /// Illegal number of encoded values
+    return {};
+  }
+
+  auto const values_count = static_cast<values_count_type>(count);
+  auto const value_bytes = bit_util::BytesForBits(value_bit_width_);
+  ARROW_DCHECK_LT(value_bytes, internal::max_size_for_v<raw_data_size_type>);
+  auto const bytes_read = header_bytes + static_cast<raw_data_size_type>(value_bytes);
+
+  auto control =
+      handler.OnRleRun(RleRun(data_ + header_bytes, values_count, value_bit_width_));
+
+  return {bytes_read, control};
+}
+
+template <typename Handler>
+void RleBitPackedParser::Parse(Handler&& handler) {
+  while (!Exhausted()) {
+    auto [read, control] = PeekImpl(handler);
+    data_ += read;
+    data_size_ -= read;
+    if (ARROW_PREDICT_FALSE(control == ControlFlow::Break)) {
+      break;
     }
-    total_processed += processed;
-    out += block.length;
-    valid_bits_offset += block.length;
-  } while (processed == block.length);
-  return total_processed;
+  }
 }
 
+/****************
+ *  RleDecoder  *
+ ****************/
+
 template <typename T>
-bool RleDecoder::NextCounts() {
-  // Read the next run's indicator int, it could be a literal or repeated run.
-  // The int is encoded as a vlq-encoded value.
-  uint32_t indicator_value = 0;
-  if (!bit_reader_.GetVlqInt(&indicator_value)) return false;
-
-  // lsb indicates if it is a literal run or repeated run
-  bool is_literal = indicator_value & 1;
-  uint32_t count = indicator_value >> 1;
-  if (is_literal) {
-    if (ARROW_PREDICT_FALSE(count == 0 || count > static_cast<uint32_t>(INT32_MAX) / 8)) {
-      return false;
-    }
-    literal_count_ = count * 8;
+RleDecoder<T>::RleDecoder(run_type const& run) noexcept {
```

Comment:
By the way, there seems to be no reason for these short/tiny methods to be defined outside of the class declaration, and inlining the definition there would probably make the code more readable.